### PR TITLE
[Merged by Bors] - chore(Algebra/Group/Defs): use `Nat.binaryRec`

### PR DESCRIPTION
**Pull request by astrainfinita**
---

- [x] depends on: #15567

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
We should check if the test from #17996 is impacted

File: Mathlib/Algebra/Group/Defs.lean [hunk 1]
```diff
@@ -583,40 +584,16 @@ needed. These problems do not come up in practice, so most of the time we will n
...
-  | (k + 1), y, x =>
-    let k' := (k + 1) >>> 1
-    if k &&& 1 = 1 then
-      go k' y (x + x)
-    else
-      go k' (y + x) (x + x)
+/-- Auxiliary tail-recursive implementation for `npowBinRec`. -/
+@[to_additive nsmulBinRec.go "Auxiliary tail-recursive implementation for `nsmulBinRec`."]
+abbrev npowBinRec.go {M : Type*} [Mul M] (k : ℕ) : M → M → M :=
+  k.binaryRec (fun y _ ↦ y) fun bn _n fn y x ↦ fn (cond bn (y * x) y) (x * x)
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:590]: What was the reason for moving this out of the `where` clause? This works fine:
```lean
/-- Exponentiation by repeated squaring. -/
@[to_additive "Scalar multiplication by repeated self-addition,
the additive version of exponentation by repeated squaring."]
def npowBinRec {M : Type*} [One M] [Mul M] (k : ℕ) : M → M :=
  npowBinRec.go k 1
where
  @[to_additive nsmulBinRec.go]
  go  (k : ℕ) : M → M → M :=
    k.binaryRec (fun y _ ↦ y) fun bn _n fn y x ↦ fn (cond bn (y * x) y) (x * x)
```

File: Mathlib/Algebra/Group/Defs.lean [hunk 2]
```diff
@@ -732,15 +693,15 @@ theorem nsmulRec_eq_nsmulBinRec : @nsmulRecAuto = @nsmulBinRecAuto := by
   funext M _ _ k m
   rw [nsmulBinRecAuto, nsmulRecAuto, nsmulBinRec]
   match k with
-  | 0 => rw [nsmulRec, nsmulBinRec.go]
+  | 0 => rw [nsmulRec, nsmulBinRec.go, Nat.binaryRec_zero]
   | k + 1 => rw [nsmulBinRec.go_spec, nsmulRec_eq]
 
 @[csimp]
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:699]:
```suggestion
@[to_additive (attr := csimp)]
```
and delete the lemma above.

**Comment by Vierkantor** [comment]
I'm trying to run the test from #17996 after merging the master branch, but build fails. I'll see if it's easy to fix.

File: Mathlib/Algebra/Group/Defs.lean [hunk 3]
```diff
@@ -669,41 +646,25 @@ theorem npowRec_eq {M : Type*} [Semigroup M] [One M] (k : ℕ) (m : M) :
...
-            omega
-        · omega
+  unfold go
+  generalize hk : k + 1 = k'
+  replace hk : k' ≠ 0 := by omega
+  induction k' using Nat.binaryRecFromOne generalizing n m with
+  | z₀ => simp at hk
+  | z₁ => simp [npowRec']
+  | f b k' k'0 ih =>
+    rw [Nat.binaryRec_eq rfl, ih _ _ k'0]
```
Comment (Vierkantor) [inline Mathlib/Algebra/Group/Defs.lean:663]: After merging `master`:
```suggestion
    rw [Nat.binaryRec_eq _ _ (Or.inl rfl), ih _ _ k'0]
```

**Review by Vierkantor — APPROVED** [review]
This looks like a good change to me. @eric-wieser, any further comments?

Please merge `master` into your branch and fix any build issues, then reply with `bors r+`.

bors d+

**Pull request closed without merging**